### PR TITLE
BUG: grouby(axis=1) cannot select column names

### LINKED ISSUE #27614: GroupBy(axis=1) Does Not Offer Implicit Selection By Columns Name(s)
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd
import numpy as np

df = pd.DataFrame(np.arange(12).reshape(3, 4), index=[0, 1, 0], columns=[10, 20, 10, 20])
df.index.name = "y"
df.columns.name = "x"

print df

print
print "Grouped along index:"
print df.groupby(by="y").sum()

print
print "Grouped along columns:"
# The following raises a KeyError even though  "x" is a column name
# (like "y" above, which is an index name):
df.groupby(by="x", axis=1).sum()
```
#### Problem description

The exception at the end is surprising: the intent is clearly to group by columns, on the "x" column label.

Furthermore, the documentation for `groupby()` seems to confirm this, as it states for the "by" argument that "A str or list of strs may be passed to group by the columns in ``self``".

#### Expected Output

A dataframe with index [0, 1, 0] but grouped (and summed) columns [10, 20].

I wasn't able to test with the latest Pandas version, sorry!

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 2.7.13.final.0
python-bits: 64
OS: Linux
OS-release: 2.6.32-642.15.1.el6.x86_64
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: None.None

pandas: 0.21.1
pytest: 3.2.3
pip: 9.0.1
setuptools: 28.8.0
Cython: 0.27.3
numpy: 1.13.3
scipy: 0.19.1
pyarrow: None
xarray: 0.8.2
IPython: 5.1.0
sphinx: 1.4.4
patsy: 0.4.1
dateutil: 2.6.1
pytz: 2017.2
blosc: None
bottleneck: None
tables: 3.4.4
numexpr: 2.6.5
feather: None
matplotlib: 2.1.2
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: 4.6.0
html5lib: 0.9999999
sqlalchemy: 1.2.18
pymysql: None
psycopg2: None
jinja2: 2.8
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
</details>

### PR DESCRIPTION
**Pull request by charlesdong1991**
- [x] closes #27614 
- [x] tests added / passed
- [ ] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: doc/source/whatsnew/v1.0.0.rst [hunk 1]
```diff
@@ -166,6 +166,7 @@ Groupby/resample/rolling
 
 -
 -
+- Bug in :meth:`groupby` does not offer implicit selection by column names when ``axis`` set to 1 (:issue:`27614)
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v1.0.0.rst]: Should be DataFrame.groupby`

"does not offer" to "not offering" 

"axis set to 1" to "`axis=1`"

Comment (TomAugspurger) [reply doc/source/whatsnew/v1.0.0.rst]: I'm also not sure what "implicit selection by column name" means here.

Comment (charlesdong1991) [reply doc/source/whatsnew/v1.0.0.rst]: > I'm also not sure what "implicit selection by column name" means here.

ahh, my bad, i basically just copied and pasted the issue title, will change! @TomAugspurger thanks for your quick reply!

**Review by WillAyd — CHANGES_REQUESTED** [review]
Nice looking PR! Can you add a test case for a multi index column as well?

**Comment by charlesdong1991** [comment]
@WillAyd I added tests for MI columns, and seems work well and pass all tests! Feel free to take a look again!

**Comment by jreback** [comment]
thanks @charlesdong1991

**Comment by jreback** [comment]
these *might* be closable now: https://github.com/pandas-dev/pandas/issues/21127, https://github.com/pandas-dev/pandas/issues/17961, https://github.com/pandas-dev/pandas/issues/13420, https://github.com/pandas-dev/pandas/issues/3689

**Comment by charlesdong1991** [comment]
thanks for merge!
I will double check those cases @jreback shall I create another PR to add test cases for them?

**Pull request merged**